pkg/node/node.go

```diff
@@ -474,34 +524,138 @@ func (n *Node) monitorKeepPublicKeySubmission(
-		)
-	case <-monitoringCtx.Done():
-		logger.Warningf(
-			"monitoring of public key submission for keep [%s] "+
-				"has been cancelled: [%v]",
+			"failed to get the starting block while setting up "+
+				"public key submission monitoring for keep [%s]: [%v]",
 			keepAddress.String(),
-			monitoringCtx.Err(),
+			err,
 		)
-	case <-abort:
-		logger.Warningf(
-			"monitoring of public key submission for keep [%s] "+
-				"has been aborted",
+		return
+	}
+
+	pubkeyCheckTrigger, err := n.ethereumChain.BlockCounter().BlockHeightWaiter(
+		startingBlock + blockConfirmations,
+	)
+	if err != nil {
+		logger.Errorf(
+			"failed to set up the check trigger while setting up "+
+				"public key submission monitoring for keep [%s]: [%v]",
 			keepAddress.String(),
+			err,
 		)
+		return
+	}
+
+	for {
+		select {
+		case event := <-conflictingPublicKey:
+			// Even in the case of a fork, a transaction with a conflicting
+			// public key submitted to the chain had to be signed with the
+			// operator's pubkey. For an honest operator, this situation should
+			// never happen, and if it happened, it is better to exit the
+			// monitoring immediately and do not re-attempt to submit public
+			// key for this node. It is clear that something is wrong with the
+			// operator that published the conflicting key and it is safer to
+			// abandon this keep.
+			logger.Errorf(
+				"member [%x] has submitted conflicting public key for keep [%s]: [%x]",
+				event.SubmittingMember,
+				keepAddress.String(),
+				event.ConflictingPublicKey,
+			)
+			return
+		case pubkeyCheckBlock := <-pubkeyCheckTrigger:
+			logger.Infof(
+				"checking public key submission for keep [%s] "+
+					"at block [%v]",
+				keepAddress.String(),
+				pubkeyCheckBlock,
+			)
+
+			// We check the public key periodically instead of relying on
+			// incoming events. The main motivation is that events could not be
+			// trusted because they may come from a forked chain or can
+			// the same event can be delivered multiple times.
+			keepPublicKey, err := n.ethereumChain.GetPublicKey(keepAddress)
+			if err != nil {
+				logger.Errorf(
+					"failed to get keep public key during "+
+						"public key submission monitoring for keep [%s]: [%v]",
+					keepAddress.String(),
+					err,
+				)
+			} else {
+				if len(keepPublicKey) > 0 {
+					isConfirmed, err := chainutil.WaitForBlockConfirmations(
+						n.ethereumChain.BlockCounter(),
+						pubkeyCheckBlock,
+						blockConfirmations,
+						func() (bool, error) {
+							key, err := n.ethereumChain.GetPublicKey(
+								keepAddress,
+							)
+							if err != nil {
+								return false, err
+							}
+
+							return len(key) > 0, nil
+						},
+					)
+					if err != nil {
+						logger.Errorf(
+							"failed to perform keep public key "+
+								"confirmation during public key submission "+
+								"monitoring for keep [%s]: [%v]",
+							keepAddress.String(),
+							err,
+						)
+					}
```

Comment:
Same here, I'd probably `continue` instead of submitting the key again 🤔

Reply:
We can't just `continue` because we have to set the new `pubkeyCheckTrigger` at the end of this case.

Reply:
Yeah, I see. Well, it's not the end of the world and I don't think it's worth making this code even more complex.